src/python/pants/goal/migrate_call_by_name_integration_test.py

```diff
@@ -0,0 +1,283 @@
+    from migrateme.rules1 import embedded_comments
+    from migrateme.rules1 import multiget
+    from migrateme.rules1 import multiline
+    from migrateme.rules1 import variants
+    from migrateme.rules2 import shadowed
+    from pants.engine.rules import implicitly
+    from pants.engine.goal import Goal, GoalSubsystem
+
+    from migrateme.rules1 import Bar, Baz, Foo, Thud, rules as rules1
+    from migrateme.rules2 import Qux, rules as rules2
+
+    class ContrivedGoalSubsystem(GoalSubsystem):
+        name = "contrived-goal"
+        help = "Need this to get the rule graph to create test migrations."
+
+    class ContrivedGoal(Goal):
+        subsystem_cls = ContrivedGoalSubsystem
+        environment_behavior = Goal.EnvironmentBehavior.LOCAL_ONLY
+
+    @goal_rule
+    async def setup_migrateme(black: Black) -> ContrivedGoal:
+        foo = await variants(**implicitly({black: Black}))
+        bar = await multiline(**implicitly({black: Black}))
+        baz = await embedded_comments(**implicitly({black: Black}))
+        qux = await shadowed(**implicitly({black: Black}))
+        thud = await multiget(**implicitly({black: Black}))
+
+    def rules():
+        return [*collect_rules(), *rules1(), *rules2()]
+    """
+)
+
+RULES1_FILE = dedent(
+    """\
+    from pants.backend.python.lint.black.subsystem import Black
+    from pants.backend.python.util_rules.pex import PexRequest, VenvPex
+    from pants.core.goals.check import CheckRequest, CheckResults
+    from pants.core.util_rules.archive import CreateArchive
+    from pants.core.util_rules.system_binaries import BinaryPathRequest, BinaryPaths
+    from pants.engine.environment import ChosenLocalEnvironmentName, EnvironmentName
+    from pants.engine.fs import Digest, EMPTY_SNAPSHOT
+    from pants.engine.rules import collect_rules, Get, MultiGet, rule
+    from pants.engine.target import AllTargets
+
+    class Foo:
+        pass
+
+    @rule
+    async def variants(black: Black, local_env: ChosenLocalEnvironmentName) -> Foo:
+        all_targets = await Get(AllTargets)
+        pex = await Get(VenvPex, PexRequest, black.to_pex_request())
+        digest = await Get(Digest, CreateArchive(EMPTY_SNAPSHOT))
+        paths = await Get(BinaryPaths, {{BinaryPathRequest(binary_name="time", search_path=("/usr/bin")): BinaryPathRequest, local_env.val: EnvironmentName}})
+
+    class Bar:
+        pass
+
+    @rule(desc="Ensure multi-line calls are migrated")
+    async def multiline(black: Black) -> Bar:
+        pex = await Get(
+            VenvPex,
+            PexRequest,
+            black.to_pex_request()
+        )
+
+    class Baz:
+        pass
+
+    @rule(desc="Ensure calls with embedded comments are ignored")
+    async def embedded_comments(black: Black) -> Baz:
+        pex = await Get(
+            VenvPex,
+            PexRequest,
+            # Some comment that the AST parse wipes out, so we can't migrate this call safely
```

Comment:
Embedded comments are wiped out by the AST. There are ways we can try to move this around at replacement time, but effort to value might not be there - especially as we ensure that the comment loses some context when we summarily move it.

I'd rather let the user know to either manually move/remove/replace the comment, or to extract the code with an embedded comment out. In the Pants repo, we have these cases specifically around Process calls with embedded Args

Reply:
Yea, that seems fine with me as long as the warning is rendered to remind people to look for the stripped comments in code review.